### PR TITLE
lib/promscrape: adds azure service discovery

### PR DESCRIPTION
**Pull request by f41gh7**
Adds azure service discovery mechanism
implements authorization with oauth and msi
lists virtual machines and virtual machines managed by scaleSet

https://github.com/VictoriaMetrics/VictoriaMetrics/issues/1364

### CODE REVIEW

File: lib/promscrape/discovery/azure/api.go [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+		return readCloudEndpointsFromFile(os.Getenv("AZURE_ENVIRONMENT_FILEPATH"))
+	}
+	env := cloudEnvironments[name]
+	if env == nil {
+		var supportedEnvs []string
+		for envName := range cloudEnvironments {
+			supportedEnvs = append(supportedEnvs, envName)
+		}
+
+		return nil, fmt.Errorf("incorrect value for azure env parametr: %q, supported values: %s", name, strings.Join(supportedEnvs, ","))
```
Comment (hagen1778) [inline lib/promscrape/discovery/azure/api.go:133]:
```suggestion
		return nil, fmt.Errorf("incorrect value for azure `environment` param: %q, supported values: %s", name, strings.Join(supportedEnvs, ","))
```
```suggestion
		return nil, fmt.Errorf("incorrect value for azure env parametr: %q, supported values: %s", name, strings.Join(supportedEnvs, ","))
```

File: lib/promscrape/discovery/azure/api.go [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+	}
+	return env, nil
+}
+
+func readCloudEndpointsFromFile(filePath string) (*cloudEnvironmentEndpoints, error) {
+	fileContent, err := ioutil.ReadFile(filePath)
+	if err != nil {
+		return nil, fmt.Errorf("cannot read cloud env endpoints from file: %q : %w", filePath, err)
```
Comment (hagen1778) [inline lib/promscrape/discovery/azure/api.go:141]:
```suggestion
		return nil, fmt.Errorf("cannot read cloud env endpoints from file %q:  %w", filePath, err)
```
```suggestion
		return nil, fmt.Errorf("cannot read cloud env endpoints from file: %q : %w", filePath, err)
```

File: lib/promscrape/discovery/azure/api.go [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+	}
+	var cee cloudEnvironmentEndpoints
+	if err := json.Unmarshal(fileContent, &cee); err != nil {
+		return nil, fmt.Errorf("cannot parse cloud env endpoints content: %q from file: %q: %w", string(fileContent), filePath, err)
```
Comment (hagen1778) [inline lib/promscrape/discovery/azure/api.go:145]:
```suggestion
		return nil, fmt.Errorf("cannot parse cloud env endpoints from file %q: %w", filePath, err)
```
The file content may be too verbose to print... or content sensitive info

File: lib/promscrape/discovery/azure/api.go [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+	}
+	return &cee, nil
+}
+
+func newTokenRefresher(sdc *SDConfig, ac, proxyAc *promauth.Config, cloudEndpoints *cloudEnvironmentEndpoints) (tokenRefresher, error) {
+
+	var tokenEndpoint, tokenAPIPath string
+	var modifyRequest func(request *fasthttp.Request)
+	switch sdc.AuthenticationMethod {
+	case "OAuth":
```
Comment (hagen1778) [inline lib/promscrape/discovery/azure/api.go:155]: can we have links to api specs here for each method, if available?

File: lib/promscrape/discovery/azure/api.go [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+		tokenEndpoint = cloudEndpoints.ActiveDirectoryEndpoint
+		modifyRequest = func(request *fasthttp.Request) {
+			request.SetBody([]byte(authParams))
+			request.Header.SetMethod("POST")
+		}
+	case "ManagedIdentity":
+		endpoint := resolveMSIEndpoint()
+		endpointURL, err := url.Parse(endpoint)
+		if err != nil {
+			return nil, fmt.Errorf("cannot parse MSI endpoint url: %q : %w", endpoint, err)
```
Comment (hagen1778) [inline lib/promscrape/discovery/azure/api.go:172]:
```suggestion
			return nil, fmt.Errorf("cannot parse MSI endpoint url %q: %w", endpoint, err)
```
```suggestion
			return nil, fmt.Errorf("cannot parse MSI endpoint url: %q : %w", endpoint, err)
```

File: lib/promscrape/discovery/azure/azure.go [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+	ClientSecret         promauth.Secret `yaml:"client_secret,omitempty"`
+	AuthenticationMethod string          `yaml:"authentication_method,omitempty"`
+	ResourceGroup        string          `yaml:"resource_group,omitempty"`
+
+	HTTPClientConfig  promauth.HTTPClientConfig  `yaml:",inline"`
+	ProxyURL          *proxy.URL                 `yaml:"proxy_url,omitempty"`
+	ProxyClientConfig promauth.ProxyClientConfig `yaml:",inline"`
+
+	// RefreshInterval time.Duration `yaml:"refresh_interval"`
+	// refresh_interval is obtained from `-promscrape.consulSDCheckInterval` command-line option.
```
Comment (hagen1778) [inline lib/promscrape/discovery/azure/azure.go:37]: nit: should be deleted?

File: lib/promscrape/discovery/azure/azure.go [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+	}
+	if len(sdc.Environment) == 0 {
+		sdc.Environment = "AZURECLOUD"
+	}
+	if len(sdc.AuthenticationMethod) == 0 {
+		sdc.AuthenticationMethod = "OAuth"
+	}
+	checkRequired := func(name, value string) error {
+		if len(value) == 0 {
+			return fmt.Errorf("required param for azure sd: %q is not set", name)
```
Comment (hagen1778) [inline lib/promscrape/discovery/azure/azure.go:58]:
```suggestion
			return fmt.Errorf("param %q is required for azure SD", name)
```
```suggestion
			return fmt.Errorf("required param for azure sd: %q is not set", name)
```

File: lib/promscrape/discovery/azure/azure.go [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+		}
+		if err := checkRequired("client_secret", sdc.ClientSecret.S); err != nil {
+			return err
+		}
+		if err := checkRequired("tenant_id", sdc.TenantID); err != nil {
+			return err
+		}
+	case "ManagedIdentity":
+	default:
+		return fmt.Errorf("unsupported `authentication_method`: %q for azure_sd_configs, only `OAuth` and `ManagedIdentity` supported by vmagent", sdc.AuthenticationMethod)
```
Comment (hagen1778) [inline lib/promscrape/discovery/azure/azure.go:78]:
```suggestion
		return fmt.Errorf("unsupported `authentication_method: %q` for azure SD, only `OAuth` and `ManagedIdentity` are supported", sdc.AuthenticationMethod)
```
```suggestion
		return fmt.Errorf("unsupported `authentication_method`: %q for azure_sd_configs, only `OAuth` and `ManagedIdentity` supported by vmagent", sdc.AuthenticationMethod)
```

File: lib/promscrape/discovery/azure/azure.go [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+	}
+	return nil
+}
+
+// GetLabels returns Consul labels according to sdc.
```
Comment (hagen1778) [inline lib/promscrape/discovery/azure/azure.go:83]: nit: wrong comment

File: lib/promscrape/discovery/azure/azure.go [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+			}
+			if len(sdc.TenantID) > 0 {
+				m["__meta_azure_tenant_id"] = sdc.TenantID
+			}
+			// /subscriptions/SUBSCRIPTION_ID/resourceGroups/RESOURCE_GROUP/providers/PROVIDER/TYPE/NAME
+			idPath := strings.Split(vm.ID, "/")
+			if len(idPath) > 4 {
+				m["__meta_azure_machine_resource_group"] = idPath[4]
+			}
+			if len(vm.Properties.StorageProfile.OsDisk.OsType) > 0 {
```
Comment (hagen1778) [inline lib/promscrape/discovery/azure/azure.go:126]: nit: any chance one of the objects would be nil?

File: lib/promscrape/discovery/azure/machine.go [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+	scaleSetRefs, err := listScaleSetRefs(ac)
+	if err != nil {
+		return nil, fmt.Errorf("cannot list scaleSets: %w", err)
+	}
+	ssvms, err := listScaleSetVMs(ac, scaleSetRefs)
+	if err != nil {
+		return nil, fmt.Errorf("cannot list scaleSets virtual machines: %w", err)
+	}
+	vms = append(vms, ssvms...)
+	// operations IO bound, it's ok to spawn move goroutines then CPU
```
Comment (hagen1778) [inline lib/promscrape/discovery/azure/machine.go:101]:
```suggestion
	// operations IO bound, it's ok to spawn more goroutines then CPU
```

**Review by valyala — APPROVED** [review]

**Pull request merged**